### PR TITLE
Initial work on adding support for the multiple TF Providers

### PR DESCRIPTION
**Pull request by ytimocin**
# Description

_Please explain the changes you've made._

## Type of change

- This pull request fixes a bug in Radius and has an approved issue (issue link required).
- This pull request adds or changes features of Radius and has an approved issue (issue link required).
- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Radius (issue link optional).

Fixes: #issue_number

### CODE REVIEW

File: typespec/Applications.Core/environments.tsp [hunk 1]
```diff
@@ -75,25 +76,31 @@ model EnvironmentProperties {
...
+model AuthConfig {
   @doc("Authentication information used to access private Terraform modules from Git repository sources.")
   git?: GitAuthConfig;
 }
 
 @doc("Authentication information used to access private Terraform modules from Git repository sources.")
-model GitAuthConfig{
-  @doc("Personal Access Token (PAT) configuration used to authenticate to Git platforms.")  
+model GitAuthConfig {
+  @doc("Personal Access Token (PAT) configuration used to authenticate to Git platforms..")
```
Comment (kachawla) [inline typespec/Applications.Core/environments.tsp]:
```suggestion
  @doc("Personal Access Token (PAT) configuration used to authenticate to Git platforms.")
```

File: typespec/Applications.Core/environments.tsp [hunk 2]
```diff
@@ -103,24 +110,47 @@ model SecretConfig {
   secret?: string;
 }
 
-@doc("The Cloud providers configuration")
+#suppress "@azure-tools/typespec-azure-core/bad-record-type"
```
Comment (kachawla) [inline typespec/Applications.Core/environments.tsp:113]: What's causing this to be needed?

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json [hunk 3]
```diff
@@ -3639,6 +3639,26 @@
         }
       }
     },
+    "EnvironmentVariables": {
+      "type": "object",
+      "description": "EnvironmentVariables describes structure enabling environment variables to be set.",
```
Comment (kachawla) [inline swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3625]:
```suggestion
      "description": "EnvironmentVariables describes structure enabling environment variables to be set for recipe execution.",
```

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json [hunk 4]
```diff
@@ -3639,6 +3639,26 @@
...
+      "properties": {
+        "secrets": {
+          "type": "object",
+          "description": "The provider secrets passed as environment variables.",
+          "additionalProperties": {
+            "$ref": "#/definitions/ProviderSecret"
```
Comment (kachawla) [inline swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3631]: Should we give it a more generic name rather than *Provider*Secret? The env variables are not specific to Terraform recipes and are expected to be consumed by other kinds of recipes as well where this might be used in another context with different terminology. I think this is more of a question on the [design](https://github.com/radius-project/design-notes/blob/4a768094e5541e3b1ab1716c848a06f5eb21d09f/recipe/2024-02-terraform-providers.md) than really this implementation :) @lakshmimsft, thoughts?

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json [hunk 5]
```diff
@@ -5299,6 +5361,17 @@
         "authentication": {
           "$ref": "#/definitions/AuthConfig",
           "description": "Authentication information used to access private Terraform module sources. Supported module sources: Git."
+        },
+        "providers": {
+          "type": "object",
+          "description": "Specifies the details of Terraform providers.",
```
Comment (kachawla) [inline swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:5347]:
```suggestion
          "description": "Configuration for Terraform Recipe Providers. Controls how Terraform interacts with cloud providers, SaaS providers, and other APIs. For more information, please see: https://developer.hashicorp.com/terraform/language/providers/configuration.",
```

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json [hunk 6]
```diff
@@ -3639,6 +3639,26 @@
         }
       }
     },
+    "EnvironmentVariables": {
```
Comment (kachawla) [inline swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3623]: I think we need to explicitly document the format env variable key is expected in for secret values, since it would not be obvious to someone new https://github.com/radius-project/design-notes/blob/4a768094e5541e3b1ab1716c848a06f5eb21d09f/recipe/2024-02-terraform-providers.md?plain=1#L186

File: pkg/corerp/datamodel/environment.go [hunk 7]
```diff
@@ -50,12 +50,17 @@ type EnvironmentProperties struct {
 type RecipeConfigProperties struct {
 	// Terraform specifies the terraform config properties
 	Terraform TerraformConfigProperties `json:"terraform,omitempty"`
+
+	Env EnvironmentVariables `json:"env,omitempty"`
```
Comment (lakshmimsft) [inline pkg/corerp/datamodel/environment.go]: Should we call this EnvVariables or similar? Env can be mistaken to be environment which is prevalent in codebase (apologies since i wrote the design doc :D)

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 6 above)
Comment (lakshmimsft) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3623]: It is documented both in design doc [link](https://github.com/radius-project/design-notes/blob/4a768094e5541e3b1ab1716c848a06f5eb21d09f/recipe/2024-02-terraform-providers.md?plain=1#L343) and in this file as well (ln3650, ln4463). Let me know if I didn't read the question right.

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 4 above)
Comment (lakshmimsft) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3631]: Yes, ProviderSecret is not correct. My suggestion is to rename it to RecipeSecret. open to other options.
Yes, the env variables are not specific to Terraform recipes.

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 6 above)
Comment (kachawla) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3623]: It is understandable from the example in the design doc, but I'm not sure where is it explicitly described in the typespec, could you point me add a link to the line? I think we are likely talking about different things, L3650 is ProviderSecret, I'm looking for this the description for where env variable key needs to be defined - for example `MY_ENV_VAR_2` from the design doc here: https://github.com/radius-project/design-notes/blob/4a768094e5541e3b1ab1716c848a06f5eb21d09f/recipe/2024-02-terraform-providers.md?plain=1#L259

File: pkg/corerp/datamodel/environment.go [hunk 8]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
 	// Scope is the target level for deploying the aws resources
 	Scope string `json:"scope,omitempty"`
 }
+
+// EnvironmentVariables describes structure enabling environment variables to be set.
+type EnvironmentVariables struct {
+	// Additional properties for the resource
+	AdditionalProperties map[string]any `json:"additionalProperties,omitempty"`
+
+	// Secrets values provided for the resource
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]: The comment seems off, can you update it in the context of env variables?

File: pkg/corerp/datamodel/environment.go [hunk 9]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
 	// Scope is the target level for deploying the aws resources
 	Scope string `json:"scope,omitempty"`
 }
+
+// EnvironmentVariables describes structure enabling environment variables to be set.
+type EnvironmentVariables struct {
+	// Additional properties for the resource
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]: This comment doesn't seem right either

File: pkg/corerp/datamodel/environment.go [hunk 10]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
 	// Scope is the target level for deploying the aws resources
 	Scope string `json:"scope,omitempty"`
 }
+
+// EnvironmentVariables describes structure enabling environment variables to be set.
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]: Can you please add some more context in this comment.. what are these env variables related to?

File: pkg/corerp/datamodel/environment.go [hunk 11]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
...
+	// Additional properties for the resource
+	AdditionalProperties map[string]any `json:"additionalProperties,omitempty"`
+
+	// Secrets values provided for the resource
+	Secrets map[string]ProviderSecret `json:"secrets,omitempty"`
+}
+
+// ProviderSecret specifies the secret details for the provider.
+type ProviderSecret struct {
+	// Source represents the resource id for the secret store containing credentials.
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]:
```suggestion
	// Source represents the resource id for the Application.Core/SecretStore containing credentials.
```

File: pkg/corerp/datamodel/environment.go [hunk 12]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
...
+	Source string `json:"source,omitempty"`
+
+	// Key represents the key to the secret in the secret store.
+	Key string `json:"key,omitempty"`
+}
+
+type ProviderConfigProperties struct {
+	// Additional properties for the resource
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]: Let's update this comment as well. What is a "resource" here? What are the additional properties expected to be?

File: pkg/corerp/datamodel/environment.go [hunk 13]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
...
+	AdditionalProperties map[string]any `json:"additionalProperties,omitempty"`
+
+	// Secrets values provided for the resource
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]: Same here about the comment.

File: pkg/corerp/datamodel/environment.go [hunk 14]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
...
+	Secrets map[string]ProviderSecret `json:"secrets,omitempty"`
+}
+
+// ProviderSecret specifies the secret details for the provider.
+type ProviderSecret struct {
+	// Source represents the resource id for the secret store containing credentials.
+	Source string `json:"source,omitempty"`
+
+	// Key represents the key to the secret in the secret store.
+	Key string `json:"key,omitempty"`
```
Comment (kachawla) [inline pkg/corerp/datamodel/environment.go]: I don't think we want to allow empty values for these?

File: typespec/Applications.Core/environments.tsp [hunk 15]
```diff
@@ -103,24 +110,47 @@ model SecretConfig {
...
+@doc("ProviderConfigProperties specifies provider configuration details needed for recipes.")
+model ProviderConfigProperties extends Record<unknown> {
+  @doc("The secrets for the referenced provider resource.")
+  secrets?: Record<ProviderSecret>;
+}
+
+#suppress "@azure-tools/typespec-azure-core/bad-record-type"
+@doc("EnvironmentVariables describes structure enabling environment variables to be set.")
+model EnvironmentVariables extends Record<unknown> {
```
Comment (lakshmimsft) [inline typespec/Applications.Core/environments.tsp:122]: This should be EnvironmentVariables extends Record<**string**>
we always know the type in this case as opposed to values/data in Providers block

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 6 above)
Comment (lakshmimsft) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3623]: [Ln124](https://github.com/vishwahiremat/radius/pull/1/files#diff-afd83699015b6426efe6ec70fcec90303e63689cef1b48839eb95d8435acd046R124) will denote map of string and the Secret type

File: pkg/corerp/datamodel/environment.go [hunk 16]
```diff
@@ -123,3 +128,29 @@ type ProvidersAWS struct {
 	// Scope is the target level for deploying the aws resources
 	Scope string `json:"scope,omitempty"`
 }
+
+// EnvironmentVariables describes structure enabling environment variables to be set.
+type EnvironmentVariables struct {
+	// Additional properties for the resource
+	AdditionalProperties map[string]any `json:"additionalProperties,omitempty"`
```
Comment (lakshmimsft) [inline pkg/corerp/datamodel/environment.go]: pls update to map[string]string

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 6 above)
Comment (kachawla) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3623]: > [Ln124](https://github.com/vishwahiremat/radius/pull/1/files#diff-afd83699015b6426efe6ec70fcec90303e63689cef1b48839eb95d8435acd046R124) will denote map of string and the Secret type

@lakshmimsft How does a user who is authoring bicep figures out what the "string" in this map is meant for?

File: typespec/Applications.Core/environments.tsp (hunk 15 above)
Comment (ytimocin) [reply typespec/Applications.Core/environments.tsp:122]: This is what I get when I use `Record<string>`: **Type 'Record<Applications.Core.RecipeSecret>' is not assignable to type 'string'**.

I also see that this is what we did for Extenders: https://github.com/radius-project/radius/blob/main/typespec/Applications.Core/extenders.tsp#L56.

File: pkg/corerp/datamodel/environment.go (hunk 16 above)
Comment (ytimocin) [reply pkg/corerp/datamodel/environment.go]: So this one is different I guess https://github.com/radius-project/radius/blob/main/pkg/corerp/datamodel/extender.go#L73

File: swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json (hunk 4 above)
Comment (ytimocin) [reply swagger/specification/applications/resource-manager/Applications.Core/preview/2023-10-01-preview/openapi.json:3631]: Changed it to RecipeSecret.

File: typespec/Applications.Core/environments.tsp (hunk 15 above)
Comment (lakshmimsft) [reply typespec/Applications.Core/environments.tsp:122]: @ytimocin, i'm suggesting towards being more type strict in this case (extenders and provider config need 'unknown' as their values which i dont think is the case here, let me know if i'm wrong) ..here's a draft. check if this will work.
```
model AdditionalProperties extends Record<string> {}

@doc("EnvironmentVariables describes structure enabling environment variables to be set.")
model EnvironmentVariables {
  ... AdditionalProperties;

  secrets?: Record<RecipeSecret>;
}

@doc("Specifies the secret details for the provider.")
model RecipeSecret {
  @doc("The resource id for the secret store containing credentials.")
  source: string;

  @doc("The key for the secret in the secret store.")
  key: string;
}
```

File: typespec/Applications.Core/environments.tsp (hunk 2 above)
Comment (ytimocin) [reply typespec/Applications.Core/environments.tsp:113]: This is for `Record<unknown>`. Discussed offline. Changing this to `Record<string>`.

Comment (ytimocin) [reply typespec/Applications.Core/environments.tsp:113]: Discussed offline.

File: typespec/Applications.Core/environments.tsp (hunk 15 above)
Comment (ytimocin) [reply typespec/Applications.Core/environments.tsp:122]: Discussed offline

File: pkg/corerp/datamodel/environment.go (hunk 16 above)
Comment (ytimocin) [reply pkg/corerp/datamodel/environment.go]: Discussed offline
